### PR TITLE
Allow plugin developer to override RunLoop functionality

### PR DESCRIPTION
**Pull request by rossnanop**
ping @StabbyCutyou 

When writing a messaging plugin, it became apparent that I would like to be able to control my own loop, specifically for reading messages and passing them directly to the dispatcher. To this end, I created a CustomLooper interface with a RunLoopCustom method. 

I needed to make XTrigger into an interface, so I could easily do ```looper, ok := t.(triggers.CustomLooper)```, since if t is a value, you can't go from value -> interface in Go. Open to suggestions if you don't like this idea and/or need to access fields on the new xTrigger (private type) instead of using an interface.

### CODE REVIEW

**Comment by StabbyCutyou** [comment]
I don't believe the `Trigger` structs themselves have any properties of note on them - and if they needed them, we could likely pass them into the Run call anyways, so on the face of it an interface doesn't seem wrong here.

Reviewing now.

File: templates/triggers/trigger_x_custom.template [hunk 1]
```diff
@@ -66,7 +66,7 @@ func (t *{{ .Name }}Trigger) Run(conn *connection.Connection, input *{{ .Name}}T
     
 
 // Test runs the trigger, but does not block and only runs the trigger polling one time
-func (t *{{ .Name }}Trigger) Test(conn *connection.Connection, input *{{ .Name}}TriggerInput, log plog.Logger) (*{{ .Name }}TriggerOutput, error) {
+func (t *{{ .NameCamelCase false }}Trigger) Test(conn *connection.Connection, input *{{ .Name}}TriggerInput, log plog.Logger) (*{{ .Name }}TriggerOutput, error) {
```
Comment (StabbyCutyou) [inline templates/triggers/trigger_x_custom.template]: By the time you get here, the Names are already run through a pre-processor which formats them correctly.

If you are finding that they are being emitted here incorrectly, the right approach is to fix the pre-processor, not add more function calls to the template imo.

I tried to keep the templates "dumb" with as little logic as I could manage, and do everything in the pre-processor so the compiler could catch errors where possible.

File: templates/server/http/handler_trigger_x.template [hunk 2]
```diff
@@ -97,7 +97,8 @@ func (h *{{ .Name }}TriggerHandler) ServeHTTP(w http.ResponseWriter, r *http.Req
         return
     }
 
-	trigger := triggers.{{ .Name }}Trigger{}
+	// We initialize with a nil dispatcher because we do not want to dispatch events on test
+	trigger := triggers.New{{ .Name }}Trigger(nil, response.Meta)
```
Comment (StabbyCutyou) [inline templates/server/http/handler_trigger_x.template]: Nice catch here, thank you

File: templates/triggers/trigger_x.template [hunk 3]
```diff
@@ -10,12 +10,26 @@ import (
...
 )
 
+// {{ .Name }}Trigger is a trigger interface
+type {{ .Name }}Trigger interface {
+	Send(o interface{})
+	Stop()
+	RunLoop(ctx context.Context, conn *connection.Connection, input *{{ .Name }}TriggerInput, log plog.Logger) error
+	ReadLoop(ctx context.Context, l plog.Logger) error
+    Test(conn *connection.Connection, input *{{ .Name }}TriggerInput, log plog.Logger) (*{{ .Name }}TriggerOutput, error)
+}
```
Comment (StabbyCutyou) [inline templates/triggers/trigger_x.template]: Do we absolutely need this interface? Is there any way to accomplish what you we're going for with CustomLooper, with only the CustomLooper interface?

For example, where you need to inspect the trigger to see if it has a CustomLooper method on it, could you do something like:

```go
 var ti interface{} = t
  if cl, ok := ti.(CustomLooper); ok {
    ...
  } else {
    ...
  }
```
?

Comment (StabbyCutyou) [inline templates/triggers/trigger_x.template]: Also, I know goimports formats for us, but try to keep the formatting correct in the templates as best you can - ex: your Test method is not aligned correctly.

It's a dumb nitpick I know, trust me, but I did my best to keep the templates from being piles of jumbled formats.

**Comment by rossnanop** [comment]
ping @StabbyCutyou addressed your comments.

**Comment by StabbyCutyou** [comment]
Looks good to me!

So long as it's been tested, make sure that bindata is updated and give it the ol merge-aroo!

Thanks Ross, great addition to the go sdk!

**Comment by StabbyCutyou** [comment]
Merge this por favor :)

**Comment by StabbyCutyou** [comment]
If you are certain it works ;)

**Comment by StabbyCutyou** [comment]
Talked offline, Ross has tested this and assured me it works :)

LGTM

**Pull request merged**